core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMPathfinding.kt

```diff
@@ -51,151 +92,176 @@ fun findPath(
+     * weight which takes into account the total duration of the path and the time shift at the
+     * departure (with different weights): path_duration * maxDepartureDelay + departure_time_shift.
+     *
+     * <br></br> EXAMPLE Let's assume we are trying to find a train between 9am and 10am. The
+     * maxDepartureDelay is 1 hour (3600s). Let's assume we have found two possible trains:
+     * - the first one leaves at 9:59 and lasts for 20:00 min.
+     * - the second one leaves at 9:00 and lasts for 20:01 min. As we are looking for the fastest
+     *   train, the first train should have the lightest weight, which is the case with the formula
+     *   above.
+     */
+    private fun computeTotalCostUntilEdge(edge: STDCMEdge): Double {
+        val timeEnd = edge.getApproximateTimeAtLocation(edge.length)
+        val pathDuration = timeEnd - edge.totalDepartureTimeShift
+        return pathDuration * maxDepartureDelay + edge.totalDepartureTimeShift
+    }
+
+    /**
+     * Converts the "raw" heuristics based on physical blocks, returning the most optimistic
+     * distance, into heuristics based on stdcm edges, returning the most optimistic time
+     */
+    private fun makeAStarHeuristic(
+        baseBlockHeuristics: ArrayList<AStarHeuristicId<Block>>,
+        rollingStock: RollingStock
+    ): List<AStarHeuristic<STDCMEdge, STDCMEdge>> {
+        val res = ArrayList<AStarHeuristic<STDCMEdge, STDCMEdge>>()
+        for (baseBlockHeuristic in baseBlockHeuristics) {
+            res.add(
+                AStarHeuristic { edge, offset ->
+                    val distance =
+                        baseBlockHeuristic.apply(
+                            edge.block,
+                            convertOffsetToBlock(offset, edge.envelopeStartOffset)
+                        )
+                    distance / rollingStock.maxSpeed
+                }
             )
-        val extended = infraExplorers.flatMap { extendLookaheadUntil(it, 4) }
-        for (explorer in extended) {
-            val edges =
-                STDCMEdgeBuilder(explorer, graph)
-                    .setStartTime(startTime)
-                    .setStartOffset(location.offset)
-                    .setPrevMaximumAddedDelay(maxDepartureDelay)
-                    .makeAllEdges()
-            for (edge in edges) res.add(EdgeLocation(edge, Offset(0.meters)))
         }
+        return res
+    }
+
+    /** Converts locations on a block id into a location on a STDCMGraph.Edge. */
+    private fun convertLocations(
+        graph: STDCMGraph,
+        locations: Collection<PathfindingEdgeLocationId<Block>>,
+        startTime: Double,
+        maxDepartureDelay: Double,
+        rollingStock: RollingStock,
+        endBlocks: Collection<BlockId> = setOf()
+    ): Set<STDCMEdge> {
+        val res = HashSet<STDCMEdge>()
+
+        val constraints = initConstraints(fullInfra, listOf(rollingStock))
```

Comment:
The changes from [this PR](https://github.com/OpenRailAssociation/osrd/pull/7166/) have been lost in a rebase here.

And thinking about it, the constraints should be initialized once and passed as a parameter to `convertLocations` for the cache to work properly.